### PR TITLE
So many changes I cannot even explain

### PR DESCRIPTION
**Pull request by haskellcamargo**
Forgive me

### CODE REVIEW

File: ANOTHER_TODO.md [hunk 1]
```diff
@@ -35,10 +35,35 @@
 - [ ] Implement a "type getter" for statement blocks, where we can get the return path in all paths
 - [ ] Implement ad-hoc polymorphism
 
-## Future work
+## Very important
+
+- [x] Declare operators in objects in expressions
+- [x] Declare operators in objects in types
+- [x] Access operators by `.' from objects in expressions
```
Comment (danizord) [inline ANOTHER_TODO.md]: whoops :)

**Comment by luizperes** [comment]
well, well... what have we got in here? Oh, only `101 files changed`, `108 commits`... not much, not much <3

File: bin/quack [hunk 2]
```diff
@@ -0,0 +1,3 @@
... a `boolean\', not a `%s\'','TYP190' =>'Cannot infer type of non-variable `%s\'','TYP200' =>'Expected condition %d of when expression to be a `boolean\'. Got `%s\'','TYP210' =>'All results to when expression must be of same type `%s\', but result %d is `%s\'','TYP220' =>'Expected a `number\' on operand `%s\' of range expression. Got `%s\'','TYP230' =>'No type overload for operator `%s\' on type `%s\'','TYP240' =>'Condition of ternary operator should de `boolean\'. Got `%s\'','TYP250' =>'Both sides of ternary expression must have the same type. Got `%s\' and `%s\'','TYP260' =>'`%s\' is not iterable','TYP270' =>'You forgot to set a value for immutable variable `%s\', my love','TYP280' =>'Found cyclic reference for type `%s\'','TYP290' =>'Caught declaration of free variable `%s\'','TYP300' =>'Cannot assign to variable `%s\' of type `%s\' a value of type `%s\'','TYP310' =>'I\'m sorry, but type `%s\' is not callable','TYP320' =>'Function of type `%s\' expected %d arguments, but received %d','TYP330' =>'Argument number %d of call expected to be a `%s\', but it is a `%s\'','TYP340' =>'On map, cannot match the type of %s `%s\' with `%s\'','TYP350' =>'Bad equation: `%s\' and `%s\', because:
```
Comment (luizperes) [inline bin/quack:1]: Are you sure about this file here? I don't see its usefulness once it is Quack's compact version (and Quack will change a lot, making it infeasible)

File: examples/factorial.qk [hunk 3]
```diff
@@ -1,20 +1,11 @@
-(** vim: syntax=Scala *)
+import from console
```
Comment (luizperes) [inline examples/factorial.qk]: is this already working?

File: examples/factorial.qk [hunk 4]
```diff
@@ -1,20 +1,11 @@
...
-  ^ if n = 0 then 1 else [n * fact[n - 1]]
-]
-
--: compressed
-def fact! if &(= 0) then 1 else &(* fact[&0 - 1])
+fn fact(n :: number): number :- n = 1
+    then 0
+    else n * fact(n - 1)
 
+effect main
```
Comment (luizperes) [inline examples/factorial.qk:7]: I like it! :)

File: rfc/pattern_match.qtest [hunk 5]
```diff
@@ -0,0 +1,22 @@
+%%command
+php ./src/Main.php %s
```
Comment (luizperes) [inline rfc/pattern_match.qtest:2]: do you wanna create a new repository for `rfc`?

File: src/ast/stmt/EnumStmt.php [hunk 6]
```diff
@@ -1,91 +0,0 @@
-<?php
```
Comment (luizperes) [inline src/ast/stmt/EnumStmt.php:1]: enum dies?

File: src/ast/stmt/PostConditionalStmt.php [hunk 7]
```diff
@@ -1,73 +0,0 @@
-<?php
```
Comment (luizperes) [inline src/ast/stmt/PostConditionalStmt.php:1]: what happens with `when`?

File: src/ast/stmt/SwitchStmt.php [hunk 8]
```diff
@@ -1,93 +0,0 @@
-<?php
```
Comment (luizperes) [inline src/ast/stmt/SwitchStmt.php:1]: switch will be replaced by `match`, right?

File: src/lexer/Lexer.php [hunk 9]
```diff
@@ -50,40 +50,32 @@ public function __construct($input)
...
-        $this->reserve(new Word(Tag::T_FROM, 'from'));
-        $this->reserve(new Word(Tag::T_TO, 'to'));
         $this->reserve(new Word(Tag::T_BY, 'by'));
         $this->reserve(new Word(Tag::T_WHEN, 'when'));
         $this->reserve(new Word(Tag::T_UNLESS, 'unless'));
+        $this->reserve(new Word(Tag::T_MUT, 'mut'));
+        $this->reserve(new Word(Tag::T_TYPE, 'type'));
+        $this->reserve(new Word(Tag::T_MATCH, 'match'));
+        $this->reserve(new Word(Tag::T_WITH, 'with'));
+        $this->reserve(new Word(Tag::T_DATA, 'data'));
```
Comment (luizperes) [inline src/lexer/Lexer.php:77]: what is `data`?

File: src/lexer/SymbolDecypher.php [hunk 10]
```diff
@@ -25,17 +25,15 @@ class SymbolDecypher
 {
     public static function __callStatic($method, $args)
     {
-        $context = &$args[0];
+        list ($context) = $args;
 
         switch ($method) {
             case '<':
-                return static::tryMatch($context, ['<<', '<>', '<=']);
+                return static::tryMatch($context, ['<<', '<>', '<=', '</']);
```
Comment (luizperes) [inline src/lexer/SymbolDecypher.php]: `</` ?

**Review by luizperes — CHANGES_REQUESTED** [review]
Hi @haskellcamargo, overall it is very very good! Two things that I realized, though. We are removing many characteristics of `imperative` programming (maybe we should try to find a balance between `functional` and `imperative`). Other than that, I would say to take a look at `codacy/pr` and check if there is anything else that could be improved! :)

File: bin/quack (hunk 2 above)
Comment (haskellcamargo) [reply bin/quack:1]: Yes, although we don't need to update it every time, this compiled file makes it possible installing Quack in milliseconds!

File: examples/factorial.qk (hunk 3 above)
Comment (haskellcamargo) [reply examples/factorial.qk]: No yet. This depends up on #115

File: rfc/pattern_match.qtest (hunk 5 above)
Comment (haskellcamargo) [reply rfc/pattern_match.qtest:2]: I believe a new repository would really be the right way to go.

File: src/ast/stmt/PostConditionalStmt.php (hunk 7 above)
Comment (haskellcamargo) [reply src/ast/stmt/PostConditionalStmt.php:1]: Is replaced by `match`, which is a lot more powerful.

File: src/ast/stmt/SwitchStmt.php (hunk 8 above)
Comment (haskellcamargo) [reply src/ast/stmt/SwitchStmt.php:1]: Yes!

File: src/lexer/Lexer.php (hunk 9 above)
Comment (haskellcamargo) [reply src/lexer/Lexer.php:77]: It is made for algebraic data types. This only reserves the keyword. See Haskell ADTs.

File: src/lexer/SymbolDecypher.php (hunk 10 above)
Comment (haskellcamargo) [reply src/lexer/SymbolDecypher.php]: Ops, my fault.

**Comment by haskellcamargo** [comment]
About imperative programming, it really will be supported, but it will be constructed over the functional base. We cannot keep security in a purely imperative structure, but we can if we abstract if with _functional effects_. This is a necessary step to keep the language secure.

**Comment by haskellcamargo** [comment]
About Codacy, I believe the best thing is to disable it. It is really crazy. It is indexing deleted files and things introduced years ago and that not always make sense, such as escaping string, but we are not dealing with HTML.

File: examples/factorial.qk (hunk 3 above)
Comment (joaolucasl) [reply examples/factorial.qk]: As I said in #115, the `from` should come first to help with IDE completion.

File: rfc/pattern_match.qtest (hunk 5 above)
Comment (joaolucasl) [reply rfc/pattern_match.qtest:2]: We've had this repo since 2016 😂 
https://github.com/quack/rfc

File: examples/factorial.qk (hunk 3 above)
Comment (haskellcamargo) [reply examples/factorial.qk]: @joaolucasl yeah! I forgot it here! I'm gonna fix!

Comment (joaolucasl) [reply examples/factorial.qk]: 🎉

File: examples/factorial.qk (hunk 4 above)
Comment (joaolucasl) [reply examples/factorial.qk:7]: This one is a really good idea. Explicitly declaring the effects sounds good. Not having this keyword will trigger errors when trying to acess external stuff?

File: src/ast/stmt/EnumStmt.php (hunk 6 above)
Comment (haskellcamargo) [reply src/ast/stmt/EnumStmt.php:1]: Yes, it is replaced by `type`, which, which is a lot better.

File: examples/factorial.qk (hunk 4 above)
Comment (haskellcamargo) [reply examples/factorial.qk:7]: Yes!

**Comment by haskellcamargo** [comment]
@luizperes ready for review :smile:

File: examples/factorial.qk (hunk 4 above)
Comment (luizperes) [reply examples/factorial.qk:7]: That is 10/10 <3

File: rfc/pattern_match.qtest (hunk 5 above)
Comment (luizperes) [reply rfc/pattern_match.qtest:2]: I had noooooo idea of that @joaolucasl! HAHAHAHAHHA

Comment (luizperes) [reply rfc/pattern_match.qtest:2]: Maybe we should do it in there from now on, what do you guys think?

File: src/lexer/Lexer.php (hunk 9 above)
Comment (luizperes) [reply src/lexer/Lexer.php:77]: k

**Review by luizperes — APPROVED** [review]

**Comment by luizperes** [comment]
Feel free to merge it <3

File: rfc/pattern_match.qtest (hunk 5 above)
Comment (haskellcamargo) [reply rfc/pattern_match.qtest:2]: I totally approve using `rfc` repository.

**Pull request merged**